src/Parallel/Reduction.hpp

```diff
@@ -9,36 +9,87 @@
 #include "Parallel/CharmRegistration.hpp"
 #include "Parallel/ConstGlobalCache.hpp"
 #include "Parallel/PupStlCpp11.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/Requires.hpp"
 #include "Utilities/TypeTraits.hpp"
 
 namespace Parallel {
+/// \cond
+template <class... Ts>
+struct ReductionData;
+/// \endcond
+
+namespace detail {
 /*!
  * \ingroup ParallelGroup
- * \brief Used for reducing heterogeneous collection of types in a single
- * reduction call
+ * \brief Convert a `ReductionData` to a `CkReductionMsg`. Used in custom
+ * reducers.
+ */
+template <class... Ts>
+CkReductionMsg* new_reduction_msg(
+    ReductionData<Ts...>& reduction_data) noexcept {
+  return CkReductionMsg::buildNew(static_cast<int>(reduction_data.size()),
+                                  reduction_data.packed().get());
+}
+}  // namespace detail
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief The data to be reduced, and invokables to be called whenever two
+ * reduction messages are combined and after the reduction has been completed.
  *
- * For each `ReductionData` you must write a custom reducer to reduce the data.
- * Here is an example of such a function:
+ * `InvokeCombine` is a binary invokable that maps `(T current_state, T element)
```

Comment:
Use \tparam?

Reply:
Not really sure how to do that. That would be a long `tparam`.  I find the way this is written better than what I could produce with `\tparam`.